File: app/src/main/java/com/disruption/cookcentral/ui/details/DetailsFragment.java
package com.disruption.cookcentral.ui.details;

import android.os.Bundle;
import android.text.Html;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.disruption.cookcentral.R;
import com.disruption.cookcentral.databinding.FragmentDetailsBinding;
import com.disruption.cookcentral.models.AnalyzedInstructions;
import com.disruption.cookcentral.models.Ingredients;
import com.disruption.cookcentral.models.Recipe;
import com.disruption.cookcentral.models.Steps;
import com.like.LikeButton;
import com.like.OnLikeListener;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * A simple {@link Fragment} subclass.
 */
public class DetailsFragment extends Fragment {

    private FragmentDetailsBinding mBinding;
    private LikeButton mLikeButton;
    private Recipe mRecipe;
    private DetailsViewModel mDetailsViewModel;

    public DetailsFragment() {
        // Required empty public constructor
    }

    @Override
    public View onCreateView(@NotNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        mBinding = DataBindingUtil.inflate(getLayoutInflater(), R.layout.fragment_details, container, false);
        mLikeButton = mBinding.starButton;

        if (getArguments() != null) {
            mRecipe = DetailsFragmentArgs.fromBundle(getArguments()).getRecipe();
            setUpViews(mRecipe);
            setUpIngredientsRecyclerViews(mRecipe);
            setUpStepsRecyclerViews(mRecipe);

            DetailsViewModelFactory factory = new DetailsViewModelFactory(requireActivity().getApplication());
            mDetailsViewModel = new ViewModelProvider(this, factory).get(DetailsViewModel.class);

            setUpFavsListener();
            observeLikedState();
        }

        return mBinding.getRoot();
    }

    private void setUpViews(Recipe recipe) {
        mBinding.tvRecipeTime.setText(requireContext().getString(R.string.recipe_time, recipe.getReadyInMinutes(), recipe.getServings()));
        mBinding.tvRecipeName.setText(recipe.getTitle());
        mBinding.tvRecipeInstructions.setText(Html.fromHtml(recipe.getSummary()));
        Glide.with(requireContext())
                .load(recipe.getImage())
                .centerCrop()
                .placeholder(R.drawable.image_loading_animation)
                .error(R.drawable.ic_error)
                .into(mBinding.ivRecipeImage);
    }

    private void setUpIngredientsRecyclerViews(Recipe recipe) {
        IngredientAdapter adapter = new IngredientAdapter();
        mBinding.rvIngredientsList.setAdapter(adapter);
        mBinding.rvIngredientsList.setLayoutManager(new LinearLayoutManager(requireContext(), RecyclerView.HORIZONTAL, false));
        Set<Ingredients> ingredientsSet = new HashSet<>();
        for (AnalyzedInstructions instructions : recipe.getAnalyzedInstructions()) {
            for (Steps steps : instructions.getSteps()) {
                if (steps.getIngredients() != null && !steps.getIngredients().isEmpty()) {
                    ingredientsSet.addAll(steps.getIngredients());
                }
            }
        }

        if (!ingredientsSet.isEmpty()) {
            List<Ingredients> ingredients = new ArrayList<>(ingredientsSet);
            adapter.submitList(ingredients);
        } else {
            mBinding.tvIngredientsError.setVisibility(View.VISIBLE);
        }
    }

    private void setUpStepsRecyclerViews(Recipe recipe) {
        StepAdapter adapter = new StepAdapter();
        mBinding.rvInstructionsList.setAdapter(adapter);
        mBinding.rvInstructionsList.setLayoutManager(new LinearLayoutManager(requireContext(), RecyclerView.VERTICAL, false));
        for (AnalyzedInstructions instructions : recipe.getAnalyzedInstructions()) {
            if (instructions.getSteps() != null && !instructions.getSteps().isEmpty()) {
                adapter.submitList(instructions.getSteps());
            } else {
                mBinding.tvInstructionsError.setVisibility(View.VISIBLE);
            }
        }
    }

    private void observeLikedState() {
        mDetailsViewModel.isRecipeInFavs(mRecipe.getId()).observe(this, recipe -> {
            if (recipe != null) {
                mLikeButton.setLiked(true);
            } else {
                mLikeButton.setLiked(false);
            }
        });
    }

    private void setUpFavsListener() {
        mLikeButton.setOnLikeListener(new OnLikeListener() {
            @Override
            public void liked(LikeButton likeButton) {
                mDetailsViewModel.insertRecipeToFavourites(mRecipe);
                Toast.makeText(getContext(), getString(R.string.add_to_favs), Toast.LENGTH_SHORT).show();
            }

            @Override
            public void unLiked(LikeButton likeButton) {
                mDetailsViewModel.deleteRecipeFromFavourites(mRecipe);
                Toast.makeText(getContext(), getString(R.string.remove_from_favs), Toast.LENGTH_SHORT).show();
            }
        });
    }
}

File: app/src/main/java/com/disruption/cookcentral/ui/details/DetailsFragment.kt
package com.disruption.cookcentral.ui.details

import android.os.Bundle
import android.text.Html
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.disruption.cookcentral.R
import com.disruption.cookcentral.databinding.FragmentDetailsBinding
import com.disruption.cookcentral.models.CachedRecipe
import com.disruption.cookcentral.models.Ingredients
import com.disruption.cookcentral.models.Recipe
import com.like.LikeButton
import com.like.OnLikeListener
import org.koin.android.viewmodel.ext.android.viewModel
import java.util.*

/**
 * A simple [Fragment] subclass.
 */
class DetailsFragment : Fragment() {
    private lateinit var mBinding: FragmentDetailsBinding
    private var mLikeButton: LikeButton? = null
    private var mRecipe: Recipe? = null
    private val mDetailsViewModel: DetailsViewModel by viewModel()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        mBinding = DataBindingUtil.inflate(inflater, R.layout.fragment_details, container, false)
        mLikeButton = mBinding.starButton

        if (arguments != null) {
            mRecipe = DetailsFragmentArgs.fromBundle(requireArguments()).recipe
            setUpViews(mRecipe!!)
            setUpIngredientsRecyclerViews(mRecipe!!)
            setUpStepsRecyclerViews(mRecipe!!)
            //val factory = DetailsViewModelFactory(requireActivity().application)
            //mDetailsViewModel = ViewModelProvider(this, factory).get(DetailsViewModel::class.java)
            setUpFavsListener()
            observeLikedState()
        }

        return mBinding.root
    }

    private fun setUpViews(recipe: Recipe) {
        mBinding.tvRecipeTime.text = requireContext().getString(R.string.recipe_time, recipe.readyInMinutes, recipe.servings)
        mBinding.tvRecipeName.text = recipe.title
        mBinding.tvRecipeInstructions.text = Html.fromHtml(recipe.summary)

        Glide.with(requireContext())
                .load(recipe.image)
                .centerCrop()
                .placeholder(R.drawable.image_loading_animation)
                .error(R.drawable.ic_error)
                .into(mBinding.ivRecipeImage)
    }

    private fun setUpIngredientsRecyclerViews(recipe: Recipe) {
        val adapter = IngredientAdapter()
        mBinding.rvIngredientsList.adapter = adapter
        mBinding.rvIngredientsList.layoutManager =
                LinearLayoutManager(requireContext(), RecyclerView.HORIZONTAL, false)
        val ingredientsSet: MutableSet<Ingredients> = HashSet()

        for (instructions in recipe.analyzedInstructions!!) {
            for (steps in instructions.steps!!) {
                if (steps.ingredients != null && steps.ingredients!!.isNotEmpty()) {
                    ingredientsSet.addAll(steps.ingredients!!)
                }
            }
        }

        if (ingredientsSet.isNotEmpty()) {
            val ingredients: List<Ingredients> = ArrayList(ingredientsSet)
            adapter.submitList(ingredients)
        } else {
            mBinding.tvIngredientsError.visibility = View.VISIBLE
        }
    }

    private fun setUpStepsRecyclerViews(recipe: Recipe) {
        val adapter = StepAdapter()
        mBinding.rvInstructionsList.adapter = adapter
        mBinding.rvInstructionsList.layoutManager = LinearLayoutManager(requireContext(), RecyclerView.VERTICAL, false)

        for (instructions in recipe.analyzedInstructions!!) {
            if (instructions.steps != null && instructions.steps!!.isNotEmpty()) {
                adapter.submitList(instructions.steps)
            } else {
                mBinding.tvInstructionsError.visibility = View.VISIBLE
            }
        }
    }

    private fun observeLikedState() {
        mDetailsViewModel.isRecipeInFavs(mRecipe!!.id).observe(viewLifecycleOwner, Observer { recipe: CachedRecipe? ->
            mLikeButton!!.isLiked = recipe != null
        })
    }

    private fun setUpFavsListener() {
        var cachedRecipe: CachedRecipe? = null
        if (mRecipe != null) {
            cachedRecipe = CachedRecipe()
            cachedRecipe.id = mRecipe!!.id
            cachedRecipe.image = mRecipe!!.image
            cachedRecipe.readyInMinutes = mRecipe!!.readyInMinutes
            cachedRecipe.summary = mRecipe!!.summary
            cachedRecipe.title = mRecipe!!.title
            cachedRecipe.servings = mRecipe!!.servings
        }

        val finalCachedRecipe = cachedRecipe

        mLikeButton!!.setOnLikeListener(object : OnLikeListener {
            override fun liked(likeButton: LikeButton) {
                mDetailsViewModel.insertRecipeToFavourites(finalCachedRecipe!!)
                Toast.makeText(context, getString(R.string.add_to_favs), Toast.LENGTH_SHORT).show()
            }

            override fun unLiked(likeButton: LikeButton) {
                mDetailsViewModel.deleteRecipeFromFavourites(finalCachedRecipe!!)
                Toast.makeText(context, getString(R.string.remove_from_favs), Toast.LENGTH_SHORT).show()
            }
        })
    }
}